### PR TITLE
Set up audio api

### PR DESCRIPTION
**Pull request by clare**
This is the web gui for both adding a new sound, and setting up a sound schedule.

### CODE REVIEW

File: public/javascripts/includes/timeUtil.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
+'use strict';
+
+const timeUtil = {
+  unitsMap: null,
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: Probably no need for this given that it gets set just below

File: public/javascripts/includes/timeUtil.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  let minutes = parts[2];
+  let amOrPm = parts[3];
+
+  if (amOrPm === "pm" && hours < 12) {
+    hours += 12;
+  }
+  hours = hours % 24;
+
+  return hours + ":" + minutes;
+};
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js:123]: I wonder if much of this module could be done using a date handling library (less code for us to maintain, more flexible. A good choice these days for parsing, manipulating & formatting dates is https://github.com/iamkun/dayjs

File: public/javascripts/includes/util.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
+
```
Comment (mjs) [inline public/javascripts/includes/util.js]: If this file is all about functions to support the templating infrastructure you've created, maybe call it `templateutil.js`, `tmplutil.js` or similar?

If you change the name, rename the `util` object too.

File: public/javascripts/setupAudio.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+  var form = $("form");
+  var schedule = result.schedule;
+  if (schedule && schedule.combos) {
+    util.populateElements(form, schedule);
+    for (var j = 0; j < schedule.combos.length; j++)  {
+      addNewCombo(schedule.combos[j]);
+    }
+  }
+  else {
```
Comment (mjs) [inline public/javascripts/setupAudio.js]: same line as the closing brace above?

File: public/javascripts/newAudioBaitFile.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        msg = 'Requested JSON parse failed.';
+      } else if (exception === 'timeout') {
+        msg = 'Time out error.';
+      } else if (exception === 'abort') {
+        msg = 'Ajax request aborted.';
+      } else {
+        msg = 'Uncaught Error.\n' + jqXHR.responseText;
+      }
+      alert(msg);
+    },
```
Comment (mjs) [inline public/javascripts/newAudioBaitFile.js:79]: Should this be a standard handling on all our ajax calls?

File: public/javascripts/includes/timeUtil.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  }
+  return null;
+};
+
+/**
+ *  Get the number of seconds for a value, based on the unit
+ *
+ * @param   {Number} value
+ * @param   {String} unit
+ * @param   {Object} unitValues
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: This isn't a parameter for the function.

File: public/javascripts/includes/timeUtil.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  m: 60,
+  h: 3600
+};
+
+/**
+ * Parse a timestring
+ *
+ * @param  {String} string
+ * @param  {Object} opts
+ * @return {Number}
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: mention it's in seconds?

File: public/javascripts/includes/timeUtil.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+ */
+
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: would it be better to leave these lines out to make it clearer that the docstring goes with the function?
(same throughout)

File: public/javascripts/includes/timeUtil.js [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  s: 1,
+  m: 60,
+  h: 3600
+};
+
+/**
+ * Parse a timestring
+ *
+ * @param  {String} string
+ * @param  {Object} opts
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: This param doesn't exist

File: public/javascripts/includes/timeUtil.js [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+timeUtil.unitValues = {
+  s: 1,
+  m: 60,
+  h: 3600
+};
+
+/**
+ * Parse a timestring
+ *
+ * @param  {String} string
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: description?

File: public/javascripts/includes/timeUtil.js [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      let unit = group.match(/[a-z]+/g)[0];
+
+      totalSeconds += timeUtil.getSeconds(value, unit);
+    });
+  }
+  return totalSeconds;
+};
+
+/**
+ * Get the key for a unit
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: describe what a key is/where it's used?

File: public/javascripts/includes/timeUtil.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+ *
+ * @param   {String} unit
```
Comment (mjs) [inline public/javascripts/includes/timeUtil.js]: give examples of units?

**Review by mjs — APPROVED** [review]
Looks great. Lots of little things.

File: public/javascripts/newAudioBaitFile.js (hunk 5 above)
Comment (clare) [reply public/javascripts/newAudioBaitFile.js:79]: Nope we need to improve the gui...

File: public/javascripts/includes/timeUtil.js (hunk 2 above)
Comment (clare) [reply public/javascripts/includes/timeUtil.js:123]: I did look for a library to start with....that's where I got the code from.   I didn't see that one though.

**Pull request merged**